lms/authentication/_helpers.py

```diff
@@ -36,20 +37,21 @@ def get_lti_user(request):
 
     :rtype: models.LTIUser
     """
-    try:
-        return LaunchParamsAuthSchema(request).lti_user()
-    except ValidationError:
-        pass
-
-    try:
-        return BearerTokenSchema(request).lti_user()
-    except ValidationError:
-        pass
-
-    try:
-        return CanvasOAuthCallbackSchema(request).lti_user()
-    except ValidationError:
-        return None
+    bearer_token_schema = BearerTokenSchema(request)
+
+    schemas = [
+        LaunchParamsAuthSchema(request).lti_user,
+        partial(bearer_token_schema.lti_user, location="headers"),
+        partial(bearer_token_schema.lti_user, location="querystring"),
+        partial(bearer_token_schema.lti_user, location="form"),
+        CanvasOAuthCallbackSchema(request).lti_user,
+    ]
+
+    for schema in schemas:
+        try:
+            return schema()
+        except ValidationError:
+            pass
```

Comment:
`BearerTokenSchema.lti_user()` used to look for the authorization param in the request headers, query string, and form body at once. It now only looks in one place at a time, and you have to tell it which place to look in each time you call it: `bearer_token_schema.lti_user(location="headers")`. So what was one call to `bearer_token_schema.lti_user()` in this function has now become three.

The `schemas` list would be nicer if `BearerTokenSchema`'s `__init__()` took the `location` arg, rather than its `lti_user()` method. Then I wouldn't need to use `functools.partial` and I could do:

```python
    schemas = [
        LaunchParamsAuthSchema(request),
        BearerTokenSchema(location="headers"),
        BearerTokenSchema(location="querystring"),
        BearerTokenSchema(location="form"),
        CanvasOAuthCallbackSchema(request),
    ]

    for schema in schemas:
        try:
            return schema.lti_user()
        except ValidationError:
            pass
```

But it doesn't make sense for `BearerTokenSchema`'s `__init__()` to take the `location` arg. The `location` arg is only needed when you call `lti_user()`. There are other places elsewhere in the code where a `BearerTokenSchema` is constructed and `lti_user()` is never called. Those places would need to pass a useless `location` arg (or `BearerTokenSchema` would need to have a useless `location` default).